k2/torch/csrc/fsa_class.cu

```diff
@@ -0,0 +1,476 @@
+  // copy tensor attributes
+  std::unordered_map<std::string, int> tensor_attrs;
+  for (const auto &fsa : srcs)
+    for (const auto &attr : fsa.tensor_attrs) ++tensor_attrs[attr.first];
+
+  std::vector<torch::Tensor> values;
+  for (const auto &attr : tensor_attrs) {
+    // skip this attribute, as it is not included in all source Fsas.
+    if (attr.second != srcs.size()) continue;
+    for (const auto &fsa : srcs) {
+      auto iter = fsa.tensor_attrs.find(attr.first);
+      K2_CHECK(iter != fsa.tensor_attrs.end());
+      values.emplace_back(iter->second);
+    }
+    torch::Tensor value = torch::cat(values, 0);
+    SetTensorAttr(attr.first, value);
+  }
+
+  // copy ragged tensor attributes
+  std::unordered_map<std::string, int> ragged_tensor_attrs;
+  for (const auto &fsa : srcs)
+    for (const auto &attr : fsa.ragged_tensor_attrs)
+      ++ragged_tensor_attrs[attr.first];
+
+  std::vector<Ragged<int32_t>> raggeds;
+  for (const auto &attr : ragged_tensor_attrs) {
+    // skip this attribute, as it is not included in all source Fsas.
+    if (attr.second != srcs.size()) continue;
+    for (const auto &fsa : srcs) {
+      auto iter = fsa.ragged_tensor_attrs.find(attr.first);
+      K2_CHECK(iter != fsa.ragged_tensor_attrs.end());
+      raggeds.emplace_back(iter->second);
+    }
+    auto value = Cat<int32_t>(/*axis*/ 0, raggeds.size(), raggeds.data(),
+                              /*merge_map*/ nullptr);
+    SetRaggedTensorAttr(attr.first, value);
+  }
+}
+
+void FsaClass::SetScores(torch::Tensor scores) {
+  K2_CHECK_EQ(scores.numel(), fsa.NumElements());
+  K2_CHECK_EQ(scores.scalar_type(), torch::kFloat32);
+  Scores().copy_(scores.detach());
+}
+
+torch::Tensor FsaClass::Scores() {
+  auto device = DeviceFromContext(fsa.Context());
+  auto scalar_type = caffe2::TypeMeta::Make<float>();
+
+  // an Arc has 4 members
+  static_assert(sizeof(Arc) == 4 * sizeof(int32_t), "");
+
+  std::vector<int64_t> sizes = {fsa.values.Dim(), 4};  // [num_rows, num_cols]
+  std::vector<int64_t> strides = {4, 1};               // in number of elements
+  auto options = torch::device(device).dtype(scalar_type);
+
+  auto tmp_scores = torch::from_blob(
+      fsa.values.Data(), sizes, strides,
+      [saved_region = fsa.values.GetRegion()](void *) {}, options);
+  return tmp_scores.index({"...", -1});
+}
+
+int32_t FsaClass::Properties() {
+  if (properties == 0) {
+    if (fsa.NumAxes() == 2) {
+      properties = GetFsaBasicProperties(fsa);
+    } else {
+      GetFsaVecBasicProperties(fsa, nullptr, &properties);
+    }
+    if (properties & kFsaPropertiesValid != kFsaPropertiesValid) {
+      K2_LOG(FATAL) << "Fsa is not valid, properties are : " << properties
+                    << " = " << PropertiesStr() << ", arcs are : " << fsa;
+    }
+  }
+  return properties;
+}
+
+std::string FsaClass::PropertiesStr() /*const*/ {
+  return FsaPropertiesAsString(Properties());
+}
+
+torch::Tensor FsaClass::Arcs() {
```

Comment:
Do we need this function in decoding, and other functions like `PropertiesStr()`, `Labels()`?